### PR TITLE
Add basic OAuth2 provider

### PR DESCRIPTION
**Pull request by dcramer**
- Supports both implicit and authorization code flows
- Refresh token and token expiration have been added
- Allows registration of OAuth applications
- Supports bypass of prompt when user has an existing authorization (containing the given scopes)

## Api Tokens

screenshot 2017-03-20 11 42 52

## Api Applications

screenshot 2017-02-24 14 46 52

screenshot 2017-02-24 14 47 21

## Authorized Applications

screenshot 2017-03-20 11 41 16

## Authorize Prompt

screenshot 2017-03-20 11 45 38

### CODE REVIEW

**Comment by dcramer** [comment]
- Needs UI for a number of things.
- Needs to better handle the "you're not logged in" flow.

I think I have full branch coverage for the flows

**Comment by dcramer** [comment]
I think most UI is in place now, but I want to improve the unauthenticated experience yet.

**Comment by dcramer** [comment]
@MeredithAnya fyi i think i need to finish this for zendesk

File: src/sentry/static/sentry/app/views/apiApplicationDetails.jsx [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+          });
+        },
+        complete: () => {
+          IndicatorStore.remove(loadingIndicator);
+        }
+      });
+    });
+  },
+
+  onRemoveApplication(app) {
```
Comment (dcramer) [inline src/sentry/static/sentry/app/views/apiApplicationDetails.jsx]: TODO

File: src/sentry/templates/sentry/oauth-authorize.html [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+      {% if existing_user.get_avatar_type == 'upload' %}
+        <img src="{% profile_photo_url existing_user.id size 36 %}">
+      {% elif existing_user.get_avatar_type == 'gravatar' %}
+        <img src="{% gravatar_url existing_user.email size 36 default 'blank' %}">
+      {% endif %}
+    </span>
+    <span class="icon-arrow-right"></span>
+    <span class="icon-arrow-left"></span>
+    <span class="avatar">
+      APP
```
Comment (dcramer) [inline src/sentry/templates/sentry/oauth-authorize.html]: TODO

**Review by dcramer — COMMENTED** [review]
- Handle remove application (UI)
- Finish approve page
- Implement error page
- API tokens need to show which app they're created onbehalf of
- need demo/qa app which implements sentry OAuth flows

**Comment by dcramer** [comment]
I guess theres a few minor things to do yet, but I also want to refactor scopes.

I want to

- deprecate member:* and just make the scope org:members
- rename delete scopes to admin

**Comment by MeredithAnya** [comment]
@dcramer yeah when you are done would like to work on getting zendesk app out, I'm not completely sure what we need to change in it but yeah

**Comment by dcramer** [comment]
will need to pair this up with https://github.com/getsentry/getsentry/pull/866

**Comment by dcramer** [comment]
Followup: 

- [ ] create token has some really awkward behavior (modal?)
- [ ] token rotation by date (e.g. 'rotate instantly' by setting a date attribute on the api specifying tokens_valid_since)

**Comment by dcramer** [comment]
All done.

**Comment by dcramer** [comment]
both github and slack use comma separated scope values -- im inclined to do the same (vs space)

**Comment by dcramer** [comment]
TODO:

>  The authorization server MUST include the HTTP "Cache-Control"
   response header field [RFC2616] with a value of "no-store" in any
   response containing tokens, credentials, or other sensitive
   information, as well as the "Pragma" response header field [RFC2616]
   with a value of "no-cache".

**Comment by dcramer** [comment]
We also need to determine how we treat re-authorization. Should we generate a new access token? Just send back the existing valid token?

**Comment by mitsuhiko** [comment]
So wrt to re-auth my understanding of OAuth 2 when I looked into it last time was this:

* you generate a new access token and refresh token on all new auths
* alternatively to a fresh auth a user can come with a refresh token which invalidates the old access token and they get a new access and refresh token

This as a response also meant that a) large amounts of access tokens can be generated (more than one per client) and b) equally many refresh tokens can be generated.

I'm looking for a best practice on oauth now since I might been wrong on this in the past.

**Comment by mitsuhiko** [comment]
However definitely refresh tokens should give you a new access token to lower the risk of access tokens leaking. They end up in all http requests so if a user loses it and we keep refreshing it, it's very bad.

**Comment by dcramer** [comment]
Ive implemented multiple api tokens per user, as well as an authorization tracker to avoid prompts when the user has already given authorization for the requested scopes

**Comment by dcramer** [comment]
One last TODO:

We should not list tokens from applications for the end-user, but rather list ApiAuthorizations. Revoking an api authorization will revoke all tokens under it.

**Comment by dcramer** [comment]
I'm thinking I'll remove authorized apps from "Auth Tokens" and under Applications we'll do "My Applications" and "Authorized Applications".

**Comment by dcramer** [comment]
done (again)

**Comment by dcramer** [comment]
requires https://github.com/getsentry/sentry/pull/5109

File: src/sentry/models/apiapplication.py [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+
+    @property
+    def is_active(self):
+        return self.status == ApiApplicationStatus.active
+
+    def is_allowed_response_type(self, value):
+        return value in ('code', 'token')
+
+    def is_valid_redirect_uri(self, value):
+        for ruri in self.redirect_uris.split('\n'):
```
Comment (mitsuhiko) [inline src/sentry/models/apiapplication.py]: I think this might be an issue. If someone has `http://example.com` set as valid redirect uri someone could use `http://example.com.mydomain.com` and get a valid redirect target.

**Review by mitsuhiko — CHANGES_REQUESTED** [review]

File: src/sentry/models/apiapplication.py (hunk 3 above)
Comment (mattrobenolt) [inline src/sentry/models/apiapplication.py]: We also have a helper here that I use from `django-sudo` that covers all the crazy edge cases:

https://github.com/mattrobenolt/django-sudo/blob/master/sudo/utils.py#L70-L108

This is used for all of our current redirects in our codebase to make sure nothing redirects away to domains it shouldn't.

Unless there's some other objection, it'd be nice to use this to cover ourselves. There's a lot of wonky attacks that can be leveraged here and I've fixed many H1 security reports because of this.

Comment (dcramer) [inline src/sentry/models/apiapplication.py]: I'm not sure that the is_safe_url does the same thing, and I wouldn't want to use code from django_sudo (unless we vendored it). redirect URIs can include subpaths/etc

Comment (mattrobenolt) [inline src/sentry/models/apiapplication.py]: It's verifying that an absolute url is within the bounds of a safe domain. So no matter how you malform a url, it can't escape the domain intended. The redirect uri is fine for that.

I use it directly from django-sudo elsewhere since I control the project, and it has 100% test coverage over all the edge cases.

But it correctly handles subpaths and all that stuff. So it _should_ fit your usecase.

Comment (dcramer) [inline src/sentry/models/apiapplication.py]: @mattrobenolt it seems to me it's only validating aginst the host, not against two URLs. Both inputs here need to be URIs, and perform correct prefix matching.

The reason I would not use django_sudo is because it is not Sentry, and it's not a framework. It's a utility that *happens* to have code that might be related (but isnt its primary function), but that doesn't mean we should be using that code. It's the same reason I wouldn't import random functions from any other library we're using unless they're part of a formal public API which that library is intended to provide.

Comment (mattrobenolt) [inline src/sentry/models/apiapplication.py]: 🤷‍♂️ just trying to help.

**Review by mattrobenolt — APPROVED** [review]
🤷‍♂️

**Review by mitsuhiko — APPROVED** [review]

**Comment by dcramer** [comment]
I had a funky issue with Percy which was caused by using ReactPage instead of GenericReactPage. Didn't hit it in QA, but probably due to combination of being on some orgs/superuser.
